tcl/org.testeditor.tcl.dsl.ide.tests/src/org/testeditor/tcl/dsl/ide/TclContentProposalProviderTest.xtend

```diff
@@ -0,0 +1,370 @@
+		]
+	}
+	
+	@Test
+	def void testComponentTemplateProposal() {
+		// given
+		var tclSnippet = '''
+			# Some
+			
+			* some step
+			Component: GreetingApplication
+			- |
+		'''.withPackage
+
+		// when
+		val proposals = tclSnippet.proposals
+
+		// then
+		proposals => [
+			expect('Stop application')
+			expect('Start application "path"')
+			expect('Is <element> visible?')
+			expect('Type "value" into <element> and wait "1"')
+			expect('Wait for "1" seconds')
+			reject('- Stop application') // proposal must NOT include '-'
+			reject('com.example.start.path') // crossreference to template variable must not show
+		]
+	}
+	
+	@Test
+	def void testComponentIncompleteTemplateProposal() {
+		// given
+		var tclSnippet = '''
+			# Some
+			
+			* some step
+			Component: GreetingApplication
+			- St|
+		'''.withPackage
+
+		// when
+		val proposals = tclSnippet.proposals
+
+		// then
+		proposals => [
+			expect('Stop application').prefix.assertEquals('St')
+			expect('Start application "path"').prefix.assertEquals('St')
+			reject('Wait for "1" seconds')
+			reject('- Stop application') // proposal must NOT include '-'
+		]
+	}
+
+	@Test
+	def void testComponentIncompleteSpacedTemplateProposal() {
+		// given
+		var tclSnippet = '''
+			# Some
+			
+			* some step
+			Component: GreetingApplication
+			- Start |
+		'''.withPackage
+
+		// when
+		val proposals = tclSnippet.proposals
+
+		// then
+		proposals => [
+			expect('application "path"').prefix.assertEquals('')
+			reject('Stop application') // proposal must NOT include templates with different prefix
+		]
+	}
+
+
+	@Test
+	def void testComponentIncompleteMultipartMultiTemplateProposal() {
+		// given
+		var tclSnippet = '''
+			# Some
+			
+			* some step
+			Component: GreetingApplication
+			- Type |
+		'''.withPackage
+
+		// when
+		val proposals = tclSnippet.proposals
+
+		// then
+		proposals => [
+			expect('boolean "true" into <element>').prefix.assertEquals('')
+			expect('"value" into <element> and wait "1"').prefix.assertEquals('')
+			expect('"value" into <element>').prefix.assertEquals('')
+			expect('confidential "param" into <element>').prefix.assertEquals('')
+			reject('Stop application') // proposal must NOT include templates with different prefix
+		]
+	}
+
+	@Test
+	def void testComponentIncompleteMultipartTemplateProposal() {
+		// given
+		var tclSnippet = '''
+			# Some
+			
+			* some step
+			Component: GreetingApplication
+			- Start app|
+		'''.withPackage
+
+		// when
+		val proposals = tclSnippet.proposals
+
+		// then
+		proposals => [
+			expect('application "path"').prefix.assertEquals('app')
+			reject('Stop application') // proposal must NOT include templates with different prefix
+		]
+	}
+
+	@Test
+	def void testComponentDenormalizedIncompleteMultipartTemplateProposal() {
+		// given
+		var tclSnippet = '''
+			# Some
+			
+			* some step
+			Component: GreetingApplication
+			      - 	Start   	 app|
```

Comment:
I was wondering what would happen if you added space(s) between "[…]app" and the caret. I suppose the proposal is expected to be empty then, right? Should that be tested, as well? (Can it be tested in this manner at all? Even the partial parser would probably face a problem in this scenario I guess.)

Reply:
this class should already be tested by another test (with prefix, after a space, e.g. testComponentIncompleteSpacedTemplateProposal).